packages/mimir/src/rules/parameter-properties.ts

```diff
@@ -0,0 +1,313 @@
+import * as ts from 'typescript';
+import { excludeDeclarationFiles, Replacement, typescriptOnly, ConfigurableRule } from '@fimbul/ymir';
+import {
+    getLineBreakStyle,
+    isCallExpression,
+    isExpressionStatement,
+    isClassDeclaration,
+    isClassExpression,
+    isConstructorDeclaration,
+    isParameterProperty,
+    isBinaryExpression,
+    isPropertyAccessExpression,
+    isIdentifier,
+    isPropertyDeclaration,
+    getPropertyName,
+    hasSideEffects,
+    isStringLiteral,
+} from 'tsutils';
+
+export interface Options {
+    mode: Mode;
+}
+
+export enum Mode {
+    Never,
+    WhenPossible,
+    Consistent,
+}
+
+const FAILURE_STRINGS = {
+    [Mode.Never]: 'Parameter properties have been disallowed.',
+    [Mode.WhenPossible]: 'Use parameter properties when possible.',
+    [Mode.Consistent]: {
+        canBeParamPropsOnly: 'All parameters can be parameter properties.',
+        cannotBeParamPropsOnly: 'Only use parameter properties if all parameters can be parameter properties.',
+    },
+};
+
+@excludeDeclarationFiles
+@typescriptOnly
+export class Rule extends ConfigurableRule<Options> {
+    protected parseOptions(options: { mode: string } | null | undefined): Options {
+        if (options) {
+            switch (options.mode) {
+                case 'consistent':
+                    return { mode: Mode.Consistent };
+                case 'when-possible':
+                    return { mode: Mode.WhenPossible };
+            }
+        }
+        return { mode: Mode.Never };
+    }
+
+    public apply() {
+        for (const node of this.context.getFlatAst()) {
+            if (
+                (isClassDeclaration(node) || isClassExpression(node)) &&
+                node.members.filter(isConstructorDeclaration).length > 0
```

Comment:
this whole condition seems redundant. you can test directly if a node is a ConstructorDeclaration with body, no need to test for a ClassLikeDeclaration first